BUG: Fix reference counting for subarrays containing objects

seberg opened the pull request:
This fixes reference counting for subarrays containing objects. Note that the 3rd commit is currently a WIP and mostly a proof that fixing up assignment like this works to not lose references. It may well approximate the correct solution, but I am not 100% sure.

This PR contains changes moved from gh-12624

seberg commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -464,6 +464,62 @@ _strided_to_strided_contig_align_wrap_init_dest(char *dst, npy_intp dst_stride,
+            *fromdata = d->fromdata;
+    char *bufferin = d->bufferin, *bufferout = d->bufferout;
+    PyArray_StridedUnaryOp *_strided_to_strided = PyArray_GetStridedCopyFn(
+            0, dst_stride, sizeof(PyObject *), sizeof(PyObject *));
```
This is the actual change/hack. Instead of just using `memset` to initialize the buffer, copy the data from the original array into the buffer, so that decref'ing can occur. I could imagine that we can go somewhere earlier to a slow direct cast loop, which calls the cast function for every single element instead of attempting to speed it up by doing it in chunks (frankly, I doubt that helps for objects that need casting anyway).

mhvk commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -51,6 +51,31 @@ PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
+        }
+        else {
+            for (i = 0; i < size; i++){
+                PyArray_Item_INCREF(data + i * inner_elsize,
```
OK, I think I finally understand what this is actually doing - it just checks whether any part is an object and then increases its refcount. It might not be bad to write a comment to state that, so that the next person looking at the code doesn't have to look up what `PyArray_Item_INCREF` actually does - the name is not too helpful here.

Separately, the check for object is very early on; I don't think there is much to gain by making a fast path, especially since anything to do with objects is slow anyway.

seberg replied:
I was not 100% sure where you were going. Added some documentation in a new commit though.

mhvk commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -87,6 +112,31 @@ PyArray_Item_XDECREF(char *data, PyArray_Descr *descr)
+        size = descr->elsize / inner_elsize;
+
+        if (descr->subarray->base->type_num == NPY_OBJECT) {
+            /* Fast path for only-objects */
```
As above, not clear to me this is worth the extra complexity.

mhvk commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -258,6 +308,10 @@ _fillobject(char *optr, PyObject *obj, PyArray_Descr *dtype)
         }
     }
+    if (dtype->type_num == NPY_OBJECT) {
+        Py_XINCREF(obj);
```
Surely `obj` cannot be `NULL` here?

mattip replied:
@seberg do you want to change this?

seberg replied:
@mattip, I don't mind changing this. I am just not sure how much we implicitly support this, since NULL is also always checked e.g. in the ufunc loops that you worked on with the exception chaining.
But, I am pretty ambivalent about it. If you/anyone thinks we should just move away from it, lets do that. I do not really see the point in it.

mhvk commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -274,15 +328,32 @@ _fillobject(char *optr, PyObject *obj, PyArray_Descr *dtype)
+        if (dtype->subarray->base->type_num == NPY_OBJECT) {
+            /* fast path when subarray only includes objects */
+            for (i = 0; i < size; i++){
+                Py_XINCREF(obj);
```
Same, surely `obj` cannot be `NULL` (realize you probably copied from above, but it seems very odd).

seberg replied:
I think you are right in principle. But IIRC quite a few, maybe even all, internal functions actually accept interpreting `NULL` as `None` for numpy arrays. In practice I do not think numpy itself will ever fill an array with `NULL` though, so the question would be if some extensions might actually rely on being able to fill arrays with `NULL`.

mhvk commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -274,15 +328,32 @@ _fillobject(char *optr, PyObject *obj, PyArray_Descr *dtype)
-            NPY_COPY_PYOBJECT_PTR(optr, &obj);
-            optr += sizeof(obj);
+        if (dtype->subarray->base->type_num == NPY_OBJECT) {
+            /* fast path when subarray only includes objects */
```
Here I'm a bit less sure, but the slow path would seem fast enough.

seberg replied:
I timed it in this one and it was quite a big difference. On the other hand, probably almost nobody actually uses these features. And as soon as you do anything else, the actual speed is probably so slow, that it simply does not matter.

So I am good with just removing the fast paths, all of this code is pretty much only used doing creation and removal anyway.

eric-wieser replied:
+1 from me on removing this fast path.

mhvk reviewed (commented):
@seberg - could to separate it out, though looking at it again and bothering to look up `PyArray_Item_INCREF` I see that what you had made sense already.

Comments mostly about whether there is really a need for a fast path - objects are slow generally and it just adds to the complexity.

ahaldane commented:
LGMT.

I looked through the other functions calling `memset` in that file. They all seem to have an alternate 'refcnt' version already, though it's a little confusing to work through all the code-paths. Only the one you found seemed to be missing a refcnt version.

For the object fast-path, I'm happy leaving it in, though I don't think it's really needed. People using object arrays are used to them being slow already.

eric-wieser commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -51,6 +51,31 @@ PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
+        size = descr->elsize / inner_elsize;
+
+        if (descr->subarray->base->type_num == NPY_OBJECT) {
+            /* Fast path for only-objects */
```
This path isn't used in `PyDataType_HASFIELDS` - I don't see why we should optimize for it here either.

seberg commented:
OK, will remove the fast-paths again then.

@ahaldane, thanks for looking through it. I am actually only half sure what that memsetting is good for, but I guess it is probably also used for setting VOID types only partially or so.

So I guess this will be roughly right, except I should make the actual copying over code nicer (or maybe I can find a DECREF'ing copy function from buffer to output array to replace the current one).

eric-wieser commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -464,6 +464,62 @@ _strided_to_strided_contig_align_wrap_init_dest(char *dst, npy_intp dst_stride,
+    PyArray_StridedUnaryOp *_strided_to_strided = PyArray_GetStridedCopyFn(
+            0, dst_stride, sizeof(PyObject *), sizeof(PyObject *));
+
+    for(;;) {
```
I think that
```C
for(;;) {
   if (x) {
       y();
   }
   else {
      z();
   }
}
```
is clearer as
```C
while (x()) {
   y();
}
z();
```

eric-wieser commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -51,6 +51,31 @@ PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
         }
     }
+    else if (PyDataType_HASSUBARRAY(descr)) {
+        int size, i, inner_elsize;
```
Should any of these be `intp`?

seberg replied:
I think `int` is correct, since that is what dtypes use for the `elsize` so we are stuck with it.

seberg replied:
I guess we could use `intp` anyway...

eric-wieser commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -51,6 +51,31 @@ PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
+
+        inner_elsize = descr->subarray->base->elsize;
+        /* Subarrays are always contiguous in memory */
+        size = descr->elsize / inner_elsize;
```
This divide by zero for `np.dtype(('V0', (1, 2)))`

eric-wieser replied:
Hmm, I can't work out how to construct that dtype any more

eric-wieser replied:
Numpy tries really hard to stop me, but I can produce that type with:
```python
dt = np.dtype('V0')
dt = copy.copy(dt)
dt.__setstate__((3, '|', (np.dtype('V0'), (3,)), None, None, 0, 1, 0)))
```

seberg replied:
Wow, good point. Calculating from shape would be nice, but unfortunately I think the shape is only stored as a python tuple right now, so I doubt that is prettier then even checking explicitly for `inner_elsize == 0`.

I think I would just add it, with a comment that it seems currently not possible. I think the real problem would be a dtype such as `np.dtype([('a', 'i4'), ('b', [('ba', 'f', (5, 0))], (3,3))])` (a subarray containing a 0-size subarray). A simple `V0` cannot hold a referene so it can't reach this far (although that is not obvious).

eric-wieser replied:
Guess what? Numpy tries really hard to stop you producing that type too, which is clearly a bug.

eric-wieser replied:
Perhaps an intermediate step of changing to
```
#define PyDataType_ISUNSIZED(dtype) (\
    (dtype)->elsize == 0 && !PyDataType_HASSUBARRAY(dtype) && !PyDataType_HASFIELDS(dtype))
```
Would make these types creatable without going through pickle.

seberg replied:
Yeah. It is a bit strange that mine was also not creatable (not sure if you trick can manage). For here, I think I would just go with a check for `if elsize == 0: return` since it obviously cannot hold references then, and will be future proof no matter what.

I think "unsized" probably means things like `S0` which actually are used to mean "S<whatever size>"? Would have to look at the code to be quite sure.

seberg replied:
Hmmm, sorry, I guess your check would likely suffice in most places when it comes to that.

eric-wieser replied:
To be clear - I'm not saying you should use that check - I'm saying that the existing `PyDataType_ISUNSIZED` is wrong and misfires on empty subarrays and on structures with no fields.

eric-wieser replied:
> not sure if you trick can manage

Here's the general solution
```
def subarray_dtype(scalar, shape):
        """ like np.dtype((scalar, shape)), but always produces subarrays in corner cases """
	scalar = np.dtype(scalar)
	dt = np.dtype(np.void)
	dt = copy.copy(dt)
	dt.__setstate__((3, '|', (scalar, shape), None, None, scalar.itemsize * np.prod(shape), scalar.alignment, 0))
	return dt
```

eric-wieser replied:
Moved discussion about the dtype construction pain to #12681, and resolving this comment since you fixed the C code here.

seberg commented:
Pushed a different approach, which gets a function that DECREF's the original data upon copy. Better than the first hack, I think, but slightly different and it might be that the first approach was slightly better even as such. Frankly, that code is still confusing me, I think I may have to draw it out to really understand all the nesting and branching.

mhvk commented on numpy/core/tests/test_dtype.py:
```diff
@@ -445,6 +446,54 @@ def test_alignment(self):
+            # Test assumes that 0, 1, and None are singletons.
+            for func, obj in zip([np.empty, np.ones, np.zeros], [None, one, zero]):
+                if obj is None:
+                    # None test is flaky:
```
Might be good to elaborate a bit on this comment: is it numpy or python that is flaky with `None`?

seberg replied:
Ooops, leftover for some earlier version, I think the test runs fine now.

seberg replied:
Maybe I should call None also not `None` here, I got a feeling it helped for the others. I was wondering if it might have to do with the actual internal code python compiles too...

seberg replied:
Doesn't help. It all seems to have to do with trickier things. Replacing the assert with a print for example runs it fine for me.

Frankly, I am not quite sure if the other tests (one and zero) are reliable or not. They seem to be, but..

mhvk replied:
Yikes...

mhvk commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -51,11 +54,36 @@ PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
+
 /*NUMPY_API
+ *
+ * XDECREF all objects in a single array item. This is complex for
```
This is more or less what I hoped for - should help the next one looking at the code!

seberg commented:
The thing is that the function there is actually only used for
`np.empty` and `np.zeros`, so just having zeros is fine. However, maybe
I can just repeat add 100 of them to a list and allow for some
variation, we have done that in other places.

Although if these tests run fine on all python versions, I guess it is
well enough for now (and the None test is not important).

Actually, there is a test missing that I still need to add. One should
use `np.repeat`, which is one of two functions explicitly using
`PyArray_INCREF`.

Will probably get to it tomorrow. The interesting part is mostly if you
think the changes in `dtype_transfer.c` are good enough, or whether I
should unravel the logic a bit more.
It does feel a bit like it may be a band-aid solution, and I do not
like band-aid solutions. (PS: Sorry for mailing/putting it all here,
but my connection to github is seriously slow right now)

mhvk commented:
Definitely seems to be one of those "let me use this left-over hour to fix this, and o no it is a week later" kind of bugs... Maybe move the new tests to a new issue? Or is it also in this `HAS_SUBARRAY` part? (which seemed somewhat self-contained, hence the suggestion)

seberg commented:
Well, I could probably split out the `refcount.c` stuff with more limited tests for now at least.

The stuff in advanced indexing likely cooks down to changes in `dtype_transfer.c` similar to the second part here.

seberg commented:
OK, found that one hehe. Pretty obvious once you looked at the right spot. `copyswap` was just utterly broken :)

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -2236,7 +2236,9 @@ VOID_copyswapn (char *dst, npy_intp dstride, char *src, npy_intp sstride,
-    if (swap && PyArray_DESCR(arr)->subarray != NULL) {
+    if (PyDataType_REFCHK(PyArray_DESCR(arr)) ||
+                /* Cannot have objects without fields or subarrays */
+                (swap && PyArray_DESCR(arr)->subarray != NULL)) {
```
This is wrong - the below refers to `descr->subarray`, which previously guaranteed not to be null. Why is a reference check needed here?

eric-wieser replied:
I think perhaps you want `(PyDataType_REFCHK(PyArray_DESCR(arr)) || swap) && PyArray_DESCR(arr)->subarray != NULL)`.

It's not clear to me that the shortcut for `swap` makes sense in the first place, and it seems it just screws over user-defined types.

seberg replied:
Oh, moved the checks around. It is correct because of the early return in the previous if, but it reads strange like this...

seberg replied:
Correct in the sense of the inital bug. But you are right, that was wrong. Also the subarray can be of VOID type again, so then you, again, need copyswap.

Added both with comments, I think it should be right now.

eric-wieser replied:
I still think the way I wrote it makes more sense...

eric-wieser commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -258,6 +307,10 @@ _fillobject(char *optr, PyObject *obj, PyArray_Descr *dtype)
             Py_XDECREF(arr);
```
Out of scope, but - line 295 will set an exception if an integer larger than `2**64` is passed in.

seberg replied:
Frankly, I hope that this is only ever used for `0` or `None`, unfortunately it is public API.

eric-wieser commented on numpy/core/tests/test_dtype.py:
```diff
@@ -446,6 +447,149 @@ def test_alignment(self):
+        singleton = object()
+        one = 1
+        zero = 0
+        for dt, pat, count in self.iter_struct_object_dtypes(singleton):
```
It would be great to use `pytest.mark.paramtrize` here, rather than putting the for loop in the test. you might want to add a fourth `label` item to the tuple, so that you can do:
```
@pytest.mark.parametrize(['dt', 'pat', 'count', 'label'], iter_struct_object_dtypes(), id=lambda t: t[3])
def test_structured_object_create_delete(dt, pat, count, label):
    # don't actually use label here, pytest will have already handled it.
```

eric-wieser commented on numpy/core/tests/test_dtype.py:
```diff
@@ -446,6 +447,149 @@ def test_alignment(self):
+                if obj is None:
+                    # None seems flaky due to python behaviour, None's are
+                    # probably simply used in too many occasions.
+                    continue
```
With the above suggestion, this becomes `pytest.skip("None is flaky")`

eric-wieser commented on numpy/core/tests/test_dtype.py:
```diff
@@ -446,6 +447,149 @@ def test_alignment(self):
+                after_zero = sys.getrefcount(zero)
+                after_one = sys.getrefcount(one)
+                assert before_zero - after_zero == count * changed
+                assert after_one - before_one == count * changed
```
_might_ be neat to introduce a helper object for this
```
class reference_counter:
    def __init__(self, obj):
        self._obj = obj
        self._initial_count = sys.getrefcount(obj)

    @property
    def change(self):
        return sys.getrefcount(self._obj) - self._initial_count


zero_counter = reference_counter(zero)
...
assert reference_counter.change == count * changed
```

seberg replied:
Yeah, possibly. Anyway, did most (not all) of the parameterization, didn't use it much yet. If you got more, can fix it up a bit more, but probably won't continue otherwise.

Maybe should add a test for the non-contiguous subarray thing at least. But it is getting out of scope to start fixing that as well and getting late today at least.

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -2236,14 +2240,25 @@ VOID_copyswapn (char *dst, npy_intp dstride, char *src, npy_intp sstride,
+    }
+    if (swap ||
+            /* Datatype may require the copyswap function (see also else): */
+            PyDataType_HASFIELDS(descr->subarray->base) ||
```
Why are we checking for fields again, when that path was already hit above?

seberg replied:
This is on the subarray, not the original.

seberg replied:
A, maybe should include subarray in the comment...

eric-wieser replied:
Can you extract a `npy_bool can_optimize_subarray = ...` variable here, since this is really just trying to be an optimization? I think that should look like:
```
npy_bool can_optimize_subarray = !swap && !is_object && !has_fields && is_builtin;
```

eric-wieser commented on numpy/core/tests/test_dtype.py:
```diff
@@ -446,6 +447,154 @@ def test_alignment(self):
+                   ('b', [('ba', 'O'), ('bb', 'O')], (2, 3))])
+    p = (0, [[(obj, obj)] * 3] * 2)
+    yield dt, p, 12, obj, "<structured subarray 2>"
+
```
Turns out the better way to write this is `pytest.param(dt, p, 12, obj, id="<structured subarray 2>")`, which eliminate the need for the label` arguments below.

seberg replied:
oh. The pytest doc could really use a bit more example on the parameterize ref. Or maybe I just missed it.

eric-wieser replied:
No, the docs are very hard to find here.

eric-wieser commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -87,6 +115,27 @@ PyArray_Item_XDECREF(char *data, PyArray_Descr *descr)
                 PyArray_Item_XDECREF(data + offset, new);
             }
         }
```
Out of scope, but: Something weird is going on with the indent above this line

seberg commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -2236,14 +2240,27 @@ VOID_copyswapn (char *dst, npy_intp dstride, char *src, npy_intp sstride,
+        npy_bool can_optimize_subarray = (!swap &&
+                !PyDataType_HASFIELDS(descr->subarray->base) &&
+                !PyDataType_REFCHK(descr->subarray->base) &&
+                (descr->subarray->base->type_num < NPY_NTYPES));
```
Grrrrr, yeah true. Bad idea to "quickly" do things when you are already tired :) (well more generally, not this one that issue is well hidden).

eric-wieser commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -51,11 +54,36 @@ PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
+ *
+ * XDECREF all objects in a single array item. This is complex for
+ * structured datatypes where the position of objects needs to be extracted.
+ * The function calls itself recursively for complex structured dtypes.
```
Might want to avoid the word `complex` which implies `cdouble`, `csingle`. Perhaps simply "compound types"

seberg replied:
Used nested, although not sure that is the better word...

eric-wieser commented on numpy/core/tests/test_dtype.py:
```diff
@@ -446,6 +447,147 @@ def test_alignment(self):
+        # The test assumes that 0, 1, and None are singletons.
+        one = 1
+        zero = 0
+        for func, obj in zip([np.empty, np.ones, np.zeros], [None, one, zero]):
```
Why not parametrize this too:
```python
pytest.mark.parametrize(['func', 'obj'], [
    # probably simply used in too many occasions.
    pytest.param(np.empty, None, marks=pytest.mark.skip("Flaky due to python's behavior"),
    (np.ones, 1),
    (np.zeros, 0),
])
```

seberg replied:
Just a bit lazyness, done now.

seberg commented:
Added a test for the sparse field fix at least that also got in. Plus one known failure in gh-2686.

seberg commented on numpy/core/tests/test_dtype.py:
```diff
@@ -446,6 +448,172 @@ def test_alignment(self):
+    sparse_dtype = np.dtype([('a', {'names':['ab'], 'formats':['f'],
+                                    'offsets':[4]}, (2, 3))])
+
+    @pytest.mark.xfail(reason="inaccessible data is changed see gh-12686.")
```
Have to think about it. This test is good, but it spams valgrind, because uninitialized buffers are copied around. Maybe for now that is good. In the longer run – if we want to continue with such effort here – could add valgrind detection to the tests, which may be an option to skip it.

mattip replied:
Could add a `@pytest.mark.confuses_valgrind` [custom mark](https://docs.pytest.org/en/latest/example/markers.html#mark-examples), which will run the test by default, but if run as `valgrind python runtest.py -- -m "not confuses_valgrind"` will skip it

seberg replied:
Added `pytest.mark.valgrind_error` (it is a true error, not just confusion) and rebased.

charris commented:
> Oh man. I added tests, including for advanced indexing and even more bugs spawned...

That has also been my experience with adding tests :)

mattip commented:
LGTM

seberg committed:
BUG: Fix dtype object subarrays for INCREF/DECREF and zeros/empty init

seberg committed:
BUG: VOID_copyswap cannot fast path nontrivial subarrays

Especially object dtypes need reference counting, so cannot use
memcpy. Subarrays that include fields also have data which should
not be overwritten.

seberg committed:
BUG: Copying subarrays with objects must decref original objects

When not using the dtype transfer function, the buffers are initialized
to NULL, and then just copied in full. Thus, at no point decref'ing
happens on the data that is originally in the array.

seberg committed:
TST: Test subarray field reference counting

Tests initialization, copying and repeating (incrementing) as
well as advanced indexing (uses copyswap currently) and assignment

seberg committed:
TST: Test sparse subarray field assignments

Assigning to subarray fields should not change the output in
between the fields. This currently still happens for normal
assignment, but not for advanced indexing.

The second test is marked as `pytest.mark.valgrind_error` to make
it clear that it is expected to show errors in valgrind, and give
at least the theoretical the possibility to skip it during testing.

mattip commented:
Can be squash-merged (makes backporting easier) once tests pass

mattip commented:
Thanks @seberg
